L80-csharp-load-balancing.md

```diff
@@ -0,0 +1,81 @@
+    "dns://backend.default.svc.cluster.local",
+    new GrpcChannelOptions { ChannelCredentials = ChannelCredentials.Insecure });
+var client = new Greet.GreeterClient(channel);
+
+var response = await client.SayHelloAsync(new HelloRequest { Name = "world" });
+Console.WriteLine(response.Message);
+```
+
+* Address scheme will be used to create a resolver.
+  * Existing logic allows for `http` and `https`. They are special-cased to never use a resolver.
+  * A collection of resolver factories are fetched from the service provider. One is selected using the address scheme and it will create resolver. For example, `dns` will match to `DnsAddressResolverFactory` which will create a `DnsAddressResolver`.
+* Load balancer name in service configure will be used to create a load balancer.
+  * A collection of load balancer factories are fetched from the service provider. One is selected using the configured name (if none specified then `pick_first` is used) and it will create load balancer.
+  * Load balancer is initialized with an `IChannelController` instance by the channel. This is used to call back to the channel to create sub-channels and update the picker.
+
+### Connecting
+
+* No network activity until `GrpcChannel.ConnectAsync` is called, or the first RPC is started.
+* On connect, the `ConnectionManager` will subscribe to address resolver and call the resolver's `RefreshAsync` to prompt the resolver to run immediately.
+* Resolver will push addresses to `ConnectionManager` which will then call balancer.
+* Balancer will use `IChannelController` to create, remove or update sub-channels as needed based on the addresses.
+  * A new sub-channel will start connecting when `SubChannel.ConnectAsync` is called. A sub-channel will be shutdown when disposed.
+* Balancer will be notified on changes in sub-channel state. It will create a `SubChannelPicker` based on the current state.
+
+### Picking
+
+* `ConnectionManager.PickAsync` is called to get a ready sub-channel along with the sub-channel's connected address.
+* If a picker hasn't been set yet, or the picker or empty or an error then `PickAsync` will wait for a picker with a ready sub-channel. This depends on whether `WaitForReady` has been set.
+
+## Rationale
+
+* Design shares the most in common with grpc-go. SubChannels have a collection of addresses and are responsible for staying connected and retrying when in TRANSIENT_FAILURE.
+* Goal is to support basic gRPC load balancing using DNS in Kubernetes.
+* Secondary goal is to have a design in place that can be used to implement xDS in the future with minimal changes.
+* All load balancing APIs will be marked as experimental.
+* API design of resolver and load balancer purposefully choses not to use static configuration. This is considered an anti-pattern in .NET. Instead custom resolver and load balancer factories are registered with a DI container which is then hooked up to the gRPC channel using `GrpcChannelOptions.ServiceProvider`.
```

Comment:
There will probably only ever be one service provider in an app, so the effect will be the same as static, it will just require setting the ServiceProvider property on a channel when it is created. And any registrations will happen on the service provider.

Static configuration, like other gRPC libraries have, is really discouraged in .NET. Dependency injection is preferred.